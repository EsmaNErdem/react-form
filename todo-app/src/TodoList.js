import React, { useState } from "react";
import {v4 as uuid} from "uuid";
import NewTodoForm from "./NewTodoForm"
import Todo from "./Todo"

const TodoList = () => {
    const INITIAL_STATE = [
        {id: uuid(), task: "Wash dishes"}
    ]

    const [tasks, setTasks] = useState(INITIAL_STATE)

    const addTask = (newTask) => {
        setTasks(tasks => [...tasks, {...newTask, id:uuid()}])
    }

    const removeTask = (id) => {
        setTasks(tasks => tasks.filter(task => task.id !== id))
    }

    const updateTask = (id, newTask) => {
        setTasks(tasks => 
            tasks.map(task => 
                task.id === id ? {...task, task:newTask} : task
            )
        )
    }

    return (
        <div>
            <NewTodoForm addTask={addTask} />
            <ul>
                {tasks.map(({ id, task }) =>(
                    <Todo 
                    key={id} 
                    id={id} 
                    task={task} 
                    handleRemove={removeTask}
                    handleUpdate={updateTask} />
                ))}
            </ul>
        </div>
    )
}

export default TodoList;