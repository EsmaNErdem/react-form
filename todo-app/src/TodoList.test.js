import React from "react";
import { render, fireEvent, act, waitFor } from "@testing-library/react";
import TodoList from "./TodoList";

test("renders without crashing", () => {
    render(<TodoList />);
});

test("it renders and matches with snaphot", () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it("should add a new box", () => {
    const { queryByText, getByLabelText, getAllByDisplayValue } = render(<TodoList />);
    const newTaskInput = getByLabelText("New Todo Task:");
    const btn = queryByText("Add task!");

    // ensure that remove button for the box is not there yet
    expect(queryByText("Have fun")).not.toBeInTheDocument();
    fireEvent.change(newTaskInput, { target: { value: "Have fun" } });
    fireEvent.click(btn);

    // makes sure list appears in the list
    expect(queryByText("Have fun")).toBeInTheDocument();

     // expect form to be empty
    expect(newTaskInput).toHaveValue("");
    
})

test("it updates task with form", async () => {
    const { getByText, getByLabelText, queryByText, asFragment, getByDisplayValue } = render(<TodoList />);
    expect(getByText("Wash dishes")).toBeInTheDocument();
    const editBtn = getByText("Edit")
    fireEvent.click(editBtn)

    const editInput = getByLabelText("Edit Todo Task:");
    fireEvent.change(editInput, { target: {value: "No dishes today, go have fun"}})
    const updateButton = getByText("Update task!");
    fireEvent.click(updateButton)

    expect(queryByText("Wash dishes")).toBeNull();
    expect(getByText("No dishes today, go have fun")).toBeInTheDocument();

});

it("should remove a task", () => {
    const { getByText, queryByText } = render(<TodoList />);

    const removeButton = getByText('X');
    expect(removeButton).toBeInTheDocument();
    expect(getByText("Wash dishes")).toBeInTheDocument();

    fireEvent.click(removeButton);
    expect(queryByText("Wash dishes")).toBeNull();
    expect(removeButton).not.toBeInTheDocument();
});

it("should toggle a task complete", () => {
    const { getByText } = render(<TodoList />);

    const markButton = getByText('Mark Completed');
    expect(markButton).toBeInTheDocument();
    expect(getByText("Wash dishes")).not.toHaveClass("Todo-Completed");

    fireEvent.click(markButton);
    expect(getByText("Wash dishes")).toHaveClass("Todo-Completed");
    // reverse it
    fireEvent.click(markButton);
    expect(getByText("Wash dishes")).not.toHaveClass("Todo-Completed");
});