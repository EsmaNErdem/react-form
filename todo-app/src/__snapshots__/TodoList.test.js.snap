// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snaphot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="task"
      >
        New Todo Task:
      </label>
      <input
        id="task"
        name="task"
        type="text"
        value=""
      />
      <button>
        Add task!
      </button>
    </form>
    <ul>
      <div>
        <li
          class=""
        >
          Wash dishes
        </li>
        <button>
          X
        </button>
        <button>
          Edit
        </button>
        <button>
          Mark Completed
        </button>
      </div>
    </ul>
  </div>
</DocumentFragment>
`;
